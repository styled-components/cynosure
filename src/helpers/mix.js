// @flow

import endsWith from '../internalHelpers/_endsWith'
import checkFix from '../internalHelpers/_checkFix'
import Color from '../internalHelpers/_Color'
import stripUnit from '../helpers/stripUnit'

type mixColor = {
  color1: string,
  color2: string,
  weight: number,
}

/**
 * Mixes two colors together. Specifically, takes the average of each of the
 * RGB components, optionally weighted by the given percentage. The opacity
 * of the colors is also considered when weighting the components.
 *
 * The weight specifies the amount of the first color that should be included
 * in the returned color. The default, `50%`, means that half the first color
 * and half the second color should be used. `25%` means that a quarter of
 * the first color and three quarters of the second color should be used.
 *
 * @example
 * // Styles as object usage
 *
 * const styles = {
 *   ...color: mix('#f00', '#00f')
 *   ...color: mix('#f00', '#00f', 25)
 *   ...color: mix('#f00', '#00f', '25%')
 * }
 *
 *
 * // styled-components usage
 * const div = styled.div`
 *   color: ${mix('#f00', '#00f')};
 *   color: ${mix('#f00', '#00f', 25)};
 *   color: ${mix('#f00', '#00f', '25%')};
 *`
 *
 * // CSS in JS Output
 *
 * div {
 *   color: "#7f007f";
 *   color: "#3f00bf";
 *   color: "#3f00bf";
 * }
 */
function mix(color1 : mixColor | string, color2: string, weight: string|number = 50):string {
  const newColor1 = checkFix(color1)
  const newColor2 = checkFix(color2)
  let newWeight = weight
  if (typeof weight === 'string') {
    if (endsWith(weight, '%')) {
      newWeight = stripUnit(weight)// newWeight.substr(-1)
    }
  }
  if (newWeight > 100) newWeight = 100

  const c1 = new Color(newColor1)
  const c2 = new Color(newColor2)
  const newColor = new Color('mix', [c1, c2, newWeight])
  return newColor.hex()
}

export default mix
