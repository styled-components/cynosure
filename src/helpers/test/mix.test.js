// @flow
import mix from '../mix'

describe('mix', function() {
  it('should mix two colors', function() {
    expect({ color: mix('#f00', '#00f', 25) }).toMatchSnapshot()
  })

  it('should mix two colors with weight in string percentage', function() {
    expect({ color: mix('#f00', '#00f', '25%') }).toMatchSnapshot()
  })

  it('should mix two colors with default value', function() {
    expect({ color: mix('#f00', '#00f') }).toMatchSnapshot()
  })

  it('should throw an error if one color is missing', () => {
    expect(() => ({ color: mix('#f00') }))
      .toThrow('Please enter a valid colors')
  })

  it('should throw an error if empty', () => {
    expect(() => ({ color: mix() }))
      .toThrow('Please enter a valid colors')
  })

  it('should throw an error if one color is invalid', () => {
    expect(() => ({ color: mix('#test', '#f00') }))
      .toThrow('Please enter a valid colors')
  })
})
