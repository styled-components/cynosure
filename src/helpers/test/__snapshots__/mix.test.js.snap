exports[`mix should mix two colors 1`] = `
Object {
  "color": "#3f00bf",
}
`;

exports[`mix should mix two colors with default value 1`] = `
Object {
  "color": "#7f007f",
}
`;

exports[`mix should mix two colors with weight in string percentage 1`] = `
Object {
  "color": "#3f00bf",
}
`;
